contributors/design-proposals/self-hosted-kubernetes.md

```diff
@@ -0,0 +1,88 @@
+
+![](self-hosted-moving-parts.png)
+
+At the end of this process the bootkube can be shut down and the system kubelet will coordinate, through a POSIX lock, to let the self-hosted kubelet take over lifecycle and management of the control plane components. The final cluster state looks like this:
+
+![](self-hosted-final-cluster.png)
+
+
+### Bootkube Challenges
+
+This process has a number of moving parts. Most notably the hand off of control from the "host system" to the Kubernetes self-hosted system. And things can go wrong:
+
+1) The self-hosted Kubelet is in a precarious position as there is no one around to restart the process if it crashes. The high level is that the system init system will watch for the Kubelet POSIX lock and start the system Kubelet if the lock is missing. Once the system Kubelet starts it will launch the self-hosted Kubelet.
+
+2) Recovering from reboots of single-master installations is a challenge as the Kubelet won't have an API server to talk to to restart the self-hosted components. We are solving this today with "[user space checkpointing](https://github.com/kubernetes-incubator/bootkube/tree/master/cmd/checkpoint#checkpoint)" container in the Kubelet pod that will periodically check the pod manifests and persist them to the static pod manifest directory. Longer term we would like for the kubelet to be able to checkpoint itself without external code.
+
+## Long Term Goals
+
+Ideally bootkube disappears over time and is replaced by a [Kubelet pod API](https://github.com/kubernetes/kubernetes/issues/28138). The write API would enable an external installation program to setup the control plane of a self-hosted Kubernetes cluster without requiring an existing API server.
+
+[Checkpointing](https://github.com/kubernetes/kubernetes/issues/489) is also required to make for a reliable system that can survive a number of normal operations like full down scenarios of the control plane. Today, we can sufficiently do checkpointing external of the Kubelet process, but checkpointing inside of the Kubelet would be ideal.
+
+A simple updater can take care of helping users update from v1.3.0 to v1.3.1, etc over time.
+
+### Self-hosted Cluster Upgrades
+
+#### Kubelet upgrades
+
+The kubelet could be upgraded in a very similar process to that outlined in the self-hosted proposal.
+
+However, because of the challenges around the self-hosted Kubelet (see above) Tectonic has implemented an alternative scheme that side-steps the self-hosted Kubelet challenges. First, a kubelet system service is launched that uses the [chrooted kubelet](https://github.com/kubernetes/community/pull/131) implemented by the [kubelet-wrapper](https://coreos.com/kubernetes/docs/latest/kubelet-wrapper.html) then when an update is required a daemonset updates the kubelet-wrapper configuration based on version annotations and kills the kubelet PID triggering a restart.
```

Comment:
This mechanism isn't really self hosting the kubelet then?  This is a 1-4 system.  The use of DaemonSet of one-off admin actions smells a bit bad to me, to be honest.  Alternative options there would be to have the kubelet rewrite it's own config and kill itself.  This could be based off of a ConfigMap.  (Or this could be done by a KubeletConfigurer that is a true DaemonSet that watches and manages the kubelet).

The system as described here has the disadvantage that it just isn't declarative -- if a node happens to be down when an upgrade is run it'll miss out and will never be upgraded.